rs/ledger_suite/icp/ledger/src/main.rs

```diff
@@ -745,11 +748,59 @@ fn main() {
     })
 }
 
+// We use 8MiB buffer
+const BUFFER_SIZE: usize = 8388608;
+
 fn post_upgrade(args: Option<LedgerCanisterPayload>) {
     let start = dfn_core::api::performance_counter(0);
-    let mut stable_reader = stable::StableReader::new();
+
+    // In order to read the first bytes we need to use ic_cdk.
+    // dfn_core assumes the first 4 bytes store stable memory length
+    // and return bytes starting from the 5th byte.
+    let mut magic_bytes_reader = ic_cdk::api::stable::StableReader::default();
```

Comment:
To avoid possibly triggering this code without intent during a future upgrade, maybe we want to put the code that touches the memory behind a boolean that is set in the upgrade args?

Reply:
I just meant that the code is protected by a conditional like 
```
if upgrade_args.run_v1_phase{
...
}
```
This way we avoid unintentionally run the part that messes with memory. We have had instances in the past where we executed code that we did not mean to execute because of this kind of behaviour.